rpcserver.go

```diff
@@ -5397,3 +5404,149 @@ func (r *rpcServer) BakeMacaroon(ctx context.Context,
+	}
+
+	// If we did not find an open channel with the outpoint provided, check
+	// whether the channel is already closed.
+	if channelClosed {
+		channel, err := r.server.chanDB.FetchClosedChannel(outpoint)
+		// If a closed channel is not found either, return a generic channel
+		// not found error because the channel is unknown to lnd.
+		if err == channeldb.ErrClosedChannelNotFound {
+			return nil, fmt.Errorf("channel: %v not founnd", outpoint)
+		} else if err != nil {
+			return nil, err
+		}
+
+		// Set the channel ID and close height from the close summary.
+		shortChanID = channel.ShortChanID
+		closeHeight = channel.CloseHeight
+	}
+
+	// Set start time to the user specified value. If no value is provided,
+	// start time will be set to the unix epoch.
+	startTime := time.Unix(int64(req.StartTime), 0)
+
+	// Set end time from the user specified time. If no value was provided,
+	// end time will be set to the present.
+	endTime := time.Unix(int64(req.EndTime), 0)
+	if req.EndTime == 0 {
+		endTime = time.Now()
+	}
+
+	// Get open time for the channel from block height.
+	opened, err := r.getTimestampByHeight(int64(shortChanID.BlockHeight))
+	if err != nil {
+		return nil, err
+	}
+
+	// Adjust the open time backwards by the block timestamp adjustment to
+	// cover the edge case where we accidentally cut off a forward event which
+	// is timestamped before open time because the channel was opened in a
+	// block with a timestamp in the future relative to our time.
+	opened = opened.Add(blockTimestampAdjustment * -1)
+
+	// If the channel open time is after the user provided start time, increase
+	// end time to the channel open time because no revenue events would have
+	// been processed before channel open.
+	if opened.After(startTime) {
+		startTime = opened
+	}
+
+	// If the channel is closed, lookup close time for the channel.
+	if closeHeight != 0 {
+		closed, err := r.getTimestampByHeight(int64(closeHeight))
+		if err != nil {
+			return nil, err
+		}
+
+		// Adjust the close time forwards by the block timestamp adjustment to
+		// cover the edge case where we accidentally cut off a forward event which
+		// is timestamped after close time because the channel was closed in a
+		// block with a timestamp in the past relative to our time.
+		closed = closed.Add(blockTimestampAdjustment)
+
+		// If the close time on the channel is before the user provided end
+		// time, reduce to end time to channel close time because no revenue
+		// events will be processed after channel close.
+		if closed.Before(endTime) {
+			endTime = closed
+		}
+	}
+
+	// Obtain paginated forwarder events by querying the forwarder log in the
+	// period provided.
+	query := func(offset, maxEvents uint32) (channeldb.ForwardingLogTimeSlice, error) {
+		return r.server.chanDB.ForwardingLog().Query(
+			channeldb.ForwardingEventQuery{
+				StartTime:    startTime,
+				EndTime:      endTime,
+				IndexOffset:  offset,
+				NumMaxEvents: maxEvents,
+			})
```

Comment:
If you query your oldest channel and if that channel is still open, it seems that you still retrieve the complete forwarding log. I can see that for some channels the start/end range may speed up the query, but is that enough to justify the extra complexity? And is it good enough then?

I think we should first address the lower layer, the forwarding log. Make that ready for accounting and other stats like this by for example adding another index. Otherwise we are spending time on building, reviewing and maintaining code that we already know is not ideal.

Reply:
I think that these two can proceed in parallel? Reliable accounting brainstorm is on the go, and the logic of this call is unlikely to change significantly - perhaps the query changes a bit if things are better indexed, but I don't forsee it being much of an issue.

Reply:
Now that the optimization is removed, this comment does no longer apply.